Image Classification API Tutorial (#14722)
dotnet/docs: docs/machine-learning/tutorials/image-classification-api-transfer-learning.md

---
title: 'Tutorial: Automated visual inspection using transfer learning'
description: This tutorial illustrates how to use transfer learning to train a TensorFlow deep learning model in ML.NET using the image detection API to classify images of concrete surfaces as cracked or not cracked.
author: luisquintanilla
ms.author: luquinta
ms.date: 10/25/2019
ms.topic: tutorial
ms.custom: mvc
#Customer intent: As a developer, I want to use ML.NET so that I can use transfer learning in an image classification scenario to classify images using a pretrained TensorFlow model and ML.NET's Image Classification API.
---

# Tutorial: Automated visual inspection using transfer learning with the ML.NET Image Classification API

> Learn how to train a custom deep learning model using transfer learning, a pretrained TensorFlow model and the ML.NET Image Classification API to classify images of concrete surfaces as cracked or uncracked.

Reviewer: ```suggestion
Learn how to train a custom deep learning model using transfer learning, a pretrained TensorFlow model and the [ML.NET Image Classification API](URL?) to classify images of concrete surfaces as cracked or uncracked.
```

Reviewer: Provide a link for the Image Classification API? It sounds like an API separate from the ML.NET API, but I don't know where to find it.

---
title: 'Tutorial: Automated visual inspection using transfer learning'
description: This tutorial illustrates how to use transfer learning to train a TensorFlow deep learning model in ML.NET using the image detection API to classify images of concrete surfaces as cracked or not cracked.
author: luisquintanilla
ms.author: luquinta
ms.date: 10/10/2019
ms.topic: tutorial
ms.custom: mvc
#Customer intent: As a developer, I want to use ML.NET so that I can use transfer learning in an image classification scenario to classify images using a pretrained TensorFlow model and ML.NET's Image Classification API.
---

# Tutorial: Automated visual inspection using transfer learning with ML.NET's Image Classification API

> Learn to train a custom deep learning model using transfer learning and the ML.NET Image Classification API to classify images of concrete surfaces as cracked or uncracked.

Reviewer 2: Learn how to

> - Classify images with the custom model

## Image Classification Transfer Learning Sample Overview

This sample is a .NET Core console application that classifies images using a pre-trained deep learning TensorFlow model. The code for this sample can be found on the [dotnet/machinelearning-samples repository](https://github.com/dotnet/machinelearning-samples) on GitHub.

## Understand the problem

Image classification is a computer vision problem. Image classification takes an image as input and categorizes it into a prescribed class. Some scenarios where image classification is useful include:

- Facial Recognition
- Emotion detection
- Medical diagnosis
- Landmark detection

This tutorial trains a custom image classification model to perform automated visual inspection of bridge decks to identify structures that are damaged by cracks.

## ML.NET Image Classification API

ML.NET provides various ways of performing image classification. This tutorial focuses on the Image Classification API. The Image Classification API makes use of [TensorFlow.NET](https://github.com/SciSharp/TensorFlow.NET), a low-level library that provides C# bindings for the TensorFlow C++ API.

> ![ML.NET Image Classification API Architecture](./media/image-classification-api-transfer-learning/architecture.png)

Reviewer 2: This digram looks huge on my mac laptop. Possible to scale?

Author: #resolved

### What is transfer learning?

Transfer learning is the process of using knowledge gained while solving one problem and applying it to a different but related problem.

Training a deep learning model from scratch requires setting several parameters, a large amount of labeled training data and a vast amount of compute resources (hundreds of GPU hours). Using a pre-trained model along with transfer learning allows you to shortcut the training process. 

### Training process

The Image Classification API starts the training process by loading a pre-trained TensorFlow model. The training process consists of two steps:

1. Bottleneck phase
2. Training phase

![Training Steps](./media/image-classification-api-transfer-learning/training.png)

#### Bottleneck phase

> During the bottleneck phase, the set of training images are loaded and the values are used as as input, or features, for the frozen layers of the pre-trained model. The frozen layers include all of the layers in the neural network up to the penultimate layer, informally known as the bottleneck layer. These layers are referred to as frozen because no training will occur on these layers and operations are pass-through. It's at these frozen layers that the lower-level patterns that help a model differentiate between the different classes are computed. The larger the number of layers, the more computationally intensive this step is. Fortunately, since this is a one-time calculation, the results can be cached and used in later runs when experimenting with different parameters.

Reviewer 2: This is much better!

> During the bottleneck phase, the set of training images are loaded and the values are used as as input, or features, for the frozen layers of the pre-trained model. The frozen layers include all of the layers in the neural network up to the penultimate layer, informally known as the bottleneck layer. These layers are referred to as frozen because no training will occur on these layers and operations are pass-through. It's at these frozen layers that the lower-level patterns that help a model differentiate between the different classes are computed. The larger the number of layers, the more computationally intensive this step is. Fortunately, since this is a one-time calculation, the results can be cached and used in later runs when experimenting with different parameters.

Reviewer 2: Although can we be crisper about what exactly is a feature? By this definition the pixels are features?

Author: #resolved

#### Training phase

> Once the bottleneck values are computed, they are used as input to retrain the final layer of the model. This process is iterative and runs for the number of times specified by model parameters. During each run, the loss and accuracy are evaluated and the appropriate adjustments are made to improve the model with the goal of minimizing the loss and maximizing the accuracy. Once training is finished, two model formats are output. One of them is the `.pb` version of the model and the other is the `.zip` ML.NET serialized version of the model. When working in environemnts supported by ML.NET, it is recommended to use the `.zip` version of the model. However, in environemnts where ML.NET is not supported, you have the option of using the `.pb` version.

Reviewer 2: What are the bottleneck values?

Author: #resolved

> Once the output values from the bottleneck phase are computed, they are used as input to retrain the final layer of the model. This process is iterative and runs for the number of times specified by model parameters. During each run, the loss and accuracy are evaluated and the appropriate adjustments are made to improve the model with the goal of minimizing the loss and maximizing the accuracy. Once training is finished, two model formats are output. One of them is the `.pb` version of the model and the other is the `.zip` ML.NET serialized version of the model. When working in environemnts supported by ML.NET, it is recommended to use the `.zip` version of the model. However, in environemnts where ML.NET is not supported, you have the option of using the `.pb` version.

Reviewer 3: @cjgronlund it's easier to submit corrections via suggestions and in bulk to keep our inboxes sane. Did an example below.

![image](https://user-images.githubusercontent.com/12971179/67248541-ab7d8300-f419-11e9-867f-f4b2b11c5b4b.png)

```suggestion
Once the output values from the bottleneck phase are computed, they are used as input to retrain the final layer of the model. This process is iterative and runs for the number of times specified by model parameters. During each run, the loss and accuracy are evaluated and the appropriate adjustments are made to improve the model with the goal of minimizing the loss and maximizing the accuracy. Once training is finished, two model formats are output. One of them is the `.pb` version of the model and the other is the `.zip` ML.NET serialized version of the model. When working in environments supported by ML.NET, it is recommended to use the `.zip` version of the model. However, in environments where ML.NET is not supported, you have the option of using the `.pb` version.
```

Reviewer: Thanks for the tip. Sorry for the spam!

Reviewer 3: No worries. Suggestions is an awesome feature and then we can later apply them in bulk as well.

## Understand the pretrained model

> 

Reviewer 2: Should we spell out that you could classify many different classes using this model? (Feel free to steal the text from the other transfer learning tutorial (as that will probably be retired once this goes live)

Author: #resolved

    ```csharp
    var preprocessingPipeline = mlContext.Transforms.Conversion.MapValueToKey(
            inputColumnName:"Label",
            outputColumnName:"LabelAsKey")
        .Append(mlContext.Transforms.LoadImages(
            outputColumnName:"Image", 
            imageFolder: assetsRelativePath,
            useImageType: false,
            inputColumnName:"ImagePath"));
    ```

1. Use the [`Fit`](xref:Microsoft.ML.Data.EstimatorChain%601.Fit*) method to apply the data to the `preprocessingPipeline` [`EstimatorChain`](xref:Microsoft.ML.Data.EstimatorChain%601) followed by the [`Transform`](xref:Microsoft.ML.Data.TransformerChain`1.Transform*) method which returns an [`IDataView`](xref:Microsoft.ML.IDataView) containing the pre-processed data.

    ```csharp
    IDataView preProcessedData = preprocessingPipeline
                                    .Fit(shuffledData)
                                    .Transform(shuffledData);
    ```

> 1. To train a model, it's important to have a training dataset as well as a validation dataset. The model is trained on the training set of data and how well it makes predictions on unseen data is measured by the performance against the validation set. Based on the results of that performance, the model makes adustments to what it has learned in an effort to improve. The validation set can come from either splitting your original dataset or from another source that has already been set aside for this purpose. In this case, the pre-processed dataset is split into training, validation and test sets.

Reviewer 2: Suggestion:

To train a model, it's important to have a training dataset as well as a validation dataset. The model is trained on the training set. How well it makes predictions on unseen data is measured by the performance against the validation set.

Author: #resolved

> 1. To train a model, it's important to have a training dataset as well as a validation dataset. The model is trained on the training set of data and how well it makes predictions on unseen data is measured by the performance against the validation set. Based on the results of that performance, the model makes adustments to what it has learned in an effort to improve. The validation set can come from either splitting your original dataset or from another source that has already been set aside for this purpose. In this case, the pre-processed dataset is split into training, validation and test sets.

Reviewer 2: adustments -> adjustments

Author: #resolved

    ```csharp
    TrainTestData trainSplit = mlContext.Data.TrainTestSplit(data: preProcessedData, testFraction:0.3);
    TrainTestData validationTestSplit = mlContext.Data.TrainTestSplit(trainSplit.TestSet);
    ```

    The code sample above performs two splits. First, the pre-processed data is split and 70% is used for training while the remaining 30% is used for validation. Then, the 30% validation set is further split into validation and test sets where 90% is used for validation and 10% is used for testing. 

>     A way to think about the purpose of these data partitions is taking an exam. When studying for an exam, you review your notes, books or other resources to get get a grasp on the concepts that are on the exam. This is what the train set is for. Then, you might take a mock exam to validate your knowledge. This is where the validation set comes in handy. You want to check whether you have a good grasp of the concepts before taking the actual exanm. Based on those results, you take note of what you got wrong or didn't understand well and incorporate your changes as you review for the real exam. Finally, you take the exam. This is what the test set is used for. You've never seen the questions that are on the exam and now use what you learned from training and validation to apply your knowledge to the task at hand. 

Reviewer 2: exanm --> exam

Author: #resolved

> - Classify images with the custom model

## Image Classification Transfer Learning Sample Overview

This sample is a .NET Core console application that classifies images using a pre-trained deep learning TensorFlow model. The code for this sample can be found on the [dotnet/machinelearning-samples repository](https://github.com/dotnet/machinelearning-samples) on GitHub.

## Understand the problem

Image classification is a computer vision problem. Image classification takes an image as input and categorizes it into a prescribed class. Some scenarios where image classification is useful include:

- Facial Recognition
- Emotion detection
- Medical diagnosis
- Landmark detection

This tutorial trains a custom image classification model to perform automated visual inspection of bridge decks to identify structures that are damaged by cracks.

## ML.NET Image Classification API

ML.NET provides various ways of performing image classification. This tutorial focuses on the Image Classification API. The Image Classification API makes use of [TensorFlow.NET](https://github.com/SciSharp/TensorFlow.NET), a low-level library that provides C# bindings for the TensorFlow C++ API.

> ![](./media/image-classification-api-transfer-learning/architecture.png)

Reviewer 2: Do you think an ML.NET user needs to know this level of detail?

Author: #resolved

### What is transfer learning?

Transfer learning is the process of using knowledge gained while solving one problem and applying it to a different but related problem.

Training a deep learning model from scratch requires setting several parameters, a large amount of labeled training data and a vast amount of compute resources (hundreds of GPU hours). Using a pre-trained model along with transfer learning allows you to shortcut the training process. 

### Training process

The Image Classification API starts the training process by loading a pre-trained TensorFlow model. The training process consists of two steps:

1. Bottleneck phase
2. Training phase

![](./media/image-classification-api-transfer-learning/training.png)

#### Bottleneck Phase

> During the bottleneck phase, the set of training images is loaded and transformed into a format expected by the pre-trained model. These transformations include resizing the image and extracting the pixels into a tensor. A tensor can be thought of as a container that stores data in N-dimensions. Once pre-processed, the data is used as input, or features, for the frozen layers of the model. The frozen layers include all of the layers in the neural network, except the penultimate (bottleneck) layer. These layers are referred to as frozen because no training will occur on these layers and operations are pass-through. The larger the number of layers, the more computationally intensive this step is. Fortunately, since this is a one-time calculation, the results can be cached and used in later runs when experimenting with different parameters.

Reviewer 2: It's not clear to me why this is called the bottleneck phase, when the bottleneck is the penultimate layer

Author: #resolved

1. Install the **Microsoft.ML 1.4.0-preview** NuGet Package:
    1. In Solution Explorer, right-click on your project and select **Manage NuGet Packages**.
    1. Choose "nuget.org" as the Package source.
    1. Select the **Browse** tab.
    1. Check the **Include prerelease** checkbox.
    1. search for **Microsoft.ML**.
    1. Select the **Install** button.
    1. Select the **OK** button on the **Preview Changes** dialog and then select the **I Accept** button on the **License Acceptance** dialog if you agree with the license terms for the packages listed.
    1. Repeat these steps for **Microsoft.ML.Dnn 0.16.0-preview** and **Microsoft.ML.ImageAnalytics 1.4.0-preview**.

### Prepare and understand the data

> [!NOTE]
> The datasets for this tutorial are from Maguire, Marc; Dorafshan, Sattar; and Thomas, Robert J., "SDNET2018: A concrete crack image dataset for machine learning applications" (2018). Browse all Datasets. Paper 48. https://digitalcommons.usu.edu/all_datasets/48

SDNET2018 is an image dataset that contains annotations for cracked and non-cracked concrete structures (bridge decks, walls, and pavement). 

> ![](./media/image-classification-api-transfer-learning/crackeddeck.jpg)

Reviewer 2: It would be good to have a cracked and non cracked image

Author: #resolved

            if (!char.IsLetter(label[index]))
            {
                label = label.Substring(0, index);
                break;
            }
        }
    }
    ```

1. Finally, create a new instance of `ModelInput`

    ```csharp
    yield return new ModelInput()
    {
        ImagePath = file,
        Label = label
    };
    ```

### Prepare the data

1. Back in the `Main` method, use the `LoadFromDirectory` utility method to get the list of images used for training.

    ```csharp
    IEnumerable<ModelInput> images = LoadImagesFromDirectory(folder: assetsRelativePath, useFolderNameAsLabel: true);
    ```

1. Then, load the images into an [`IDataView`](xref:Microsoft.ML.IDataView) using the [`LoadFromEnumerable`](xref:Microsoft.ML.DataOperationsCatalog.LoadFromEnumerable*) method,

    ```csharp
    IDataView imageData = mlContext.Data.LoadFromEnumerable(images);
    ```

1. The way in which the data is loaded is in the order it was read from the directories. To balance the data, shuffle it using the [`ShuffleRows`](xref:Microsoft.ML.DataOperationsCatalog.ShuffleRows*) method.

    ```csharp
    IDataView shuffledData = mlContext.Data.ShuffleRows(imageData);
    ```

> 1. It's good practice to split the data into a trianing and validation set. The model is trained on the training set of data and how well it makes predictions on unseen data is measured by the performance against the validation set. Split your data 

Reviewer 2: trianing --> training

Author: #resolved

> 1. It's good practice to split the data into a trianing and validation set. The model is trained on the training set of data and how well it makes predictions on unseen data is measured by the performance against the validation set. Split your data 

Reviewer 2: I would say this in stronger terms: it's essential to have a validation set, and you can split the training data into a training and validation set, or you can get a validation set from somewhere else

Author: #resolved

    ```csharp
    TrainTestData dataSplit = mlContext.Data.TrainTestSplit(shuffledData);
    IDataView trainSet = dataSplit.TrainSet;
    IDataView validationSet = dataSplit.TestSet;
    ```

### Define the training pipeline

Model training consist of a couple of steps. First, the original text labels need to be mapped to a `KeyValue` type. Then, the Image Classification API is used to train the model.

1. In the main method, define the `MapValueToKey` estimator to map the string values in the `Label` column to the `LabelAsKey` column.

    ```csharp
    var mapLabelEstimator = mlContext.Transforms.Conversion.MapValueToKey
        (outputColumnName: "LabelAsKey",
            inputColumnName: "Label",
            keyOrdinality: ValueToKeyMappingEstimator.KeyOrdinality.ByValue);
    ```

1. Then, apply the validation set to the `mapLabelEstimator` using the [`Fit`](xref:Microsoft.ML.ConversionsExtensionsCatalog.MapValueToKey*) method, followed by the [`Transform`](xref:Microsoft.ML.ITransformer.Transform*) method to get the transformed [`IDataView`](xref:Microsoft.ML.IDataView).

    ```csharp

>     IDataView transformedValidationData = mapLabelEstimator

Reviewer 2: Should this be trainingSet?

Author: #resolved

    - `batchSize` is the number of samples to use at a time for training. During one epoch, multiple batches equal to the batchSize are used to train and update the model. The lower the number, the less memory required when each batch is processed.
    - `testOnTrainSet` tells the model to measure performance against the training set when no validation set is present.
    - `metricsCallback` binds a function to track the progress during training.
    - `validationSet` is the [`IDataView`](xref:Microsoft.ML.IDataView) containing the validation data.
    - `reuseTrainSetBottleneckCachedValues` tells the model whether to use the cached values from the bottleneck phase in subsequent runs. The bottleneck phase is a one-time pass-through computation that is computationally intensive the first time it is performed. If the training data does not change and you want to experiment using a different number of epochs or batch size, using the cached values significantly reduces the amount of time required to train a model.
    - `reuseValidationSetBottleneckCachedValues` is similar to `reuseTrainSetBottleneckCachedValues` only that in this case it's for the validation dataset.

>     - `disableEarlyStopping` tells the Image Classification whether to employ an early stopping strategy. As the model searches for the optimal values that will help it make accurate predictions during training, performance may increase or decrease. Ultimately, if the model reaches the last epoch, it may be the case that the patterns it learned from training are suboptimal. Early stopping monitors training for these drops in performance and stops the training process in an effort to preserve an optimal version of the model.

Reviewer: ```suggestion
    - `disableEarlyStopping` tells the ImageClassification estimator whether to employ an early stopping strategy. As the model searches for the optimal values that will help it make accurate predictions during training, performance may increase or decrease. Ultimately, if the model reaches the last epoch, it may be the case that the patterns it learned from training are suboptimal. Early stopping monitors training for these drops in performance and stops the training process in an effort to preserve an optimal version of the model.
```

Reviewer: Instead of "the Image Classification" did you mean "the ImageClassification estimator" that this list is about?
